main.go

```diff
@@ -76,8 +78,110 @@ func HostList(s kingpin.Settings) (target *[]string) {
 	return
 }
 
-func init() {
-	prometheus.MustRegister(versioncollector.NewCollector("smokeping_prober"))
+type smokePingers struct {
+	started     []*probing.Pinger
+	prepared    []*probing.Pinger
+	g           *errgroup.Group
+	maxInterval time.Duration
+}
+
+func (s *smokePingers) sizeOfPrepared() int {
+	if s.prepared != nil {
+		return len(s.prepared)
+	}
+	return 0
+}
+
+func (s *smokePingers) start() {
+	if s.sizeOfPrepared() == 0 {
+		return
+	}
+	if s.g != nil {
+		err := s.stop()
+		if err != nil {
+			level.Warn(logger).Log("msg", "At least one previous pinger failed to run", "err", err)
+		}
+	}
+	s.g = new(errgroup.Group)
+	s.started = s.prepared
+	splay := time.Duration(s.maxInterval.Nanoseconds() / int64(len(s.started)))
+	for _, pinger := range s.started {
+		level.Info(logger).Log("msg", "Starting prober", "address", pinger.Addr(), "interval", pinger.Interval, "size_bytes", pinger.Size, "source", pinger.Source)
+		s.g.Go(pinger.Run)
+		time.Sleep(splay)
+	}
+	s.prepared = nil
+}
+
+func (s *smokePingers) stop() error {
+	if s.g == nil {
+		return nil
+	}
+	if s.started == nil {
+		return nil
+	}
+	for _, pinger := range s.started {
+		pinger.Stop()
+	}
+	if err := s.g.Wait(); err != nil {
+		return fmt.Errorf("pingers failed: %v", err)
+	}
+	return nil
+}
+
+func (s *smokePingers) prepare(hosts *[]string, interval *time.Duration, privileged *bool, sizeBytes *int) error {
+	pingers := make([]*probing.Pinger, len(*hosts))
+	var pinger *probing.Pinger
+	var host string
+	for i, host := range *hosts {
+		pinger = probing.New(host)
+
+		err := pinger.Resolve()
+		if err != nil {
+			return fmt.Errorf("failed to resolve pinger: %v", err)
+		}
+
+		level.Info(logger).Log("msg", "Pinger resolved", "host", host, "ip_addr", pinger.IPAddr())
+
+		pinger.Interval = *interval
+		pinger.RecordRtts = false
+		pinger.SetPrivileged(*privileged)
+		pinger.Size = *sizeBytes
+
+		pingers[i] = pinger
+	}
+
+	maxInterval := *interval
+	sc.Lock()
+	for _, targetGroup := range sc.C.Targets {
+		packetSize := targetGroup.Size
+		if packetSize < 24 || packetSize > 65535 {
+			return fmt.Errorf("packet size must be in the range 24-65535, but found '%d' bytes", packetSize)
+		}
+		if targetGroup.Interval > maxInterval {
+			maxInterval = targetGroup.Interval
+		}
+		for _, host = range targetGroup.Hosts {
+			pinger = probing.New(host)
+			pinger.Interval = targetGroup.Interval
+			pinger.RecordRtts = false
+			pinger.SetNetwork(targetGroup.Network)
+			pinger.Size = packetSize
+			pinger.Source = targetGroup.Source
+			if targetGroup.Protocol == "icmp" {
+				pinger.SetPrivileged(true)
+			}
+			err := pinger.Resolve()
+			if err != nil {
+				return fmt.Errorf("failed to resolve pinger: %v", err)
+			}
+			pingers = append(pingers, pinger)
+		}
+	}
+	sc.Unlock()
```

Comment:
The `prepare` function can return (with error) before it unlocks the mutex, leaving it perpetually locked. Since `smokerpingers` is reused within the `main` function, this could result in a deadlock when the `prepare` function is called a subsequent time.

Reply:
Good catch, anyone want to open a PR for this fix?

Reply:
Good catch, indeed. I'll fix it.

Reply:
I opened PR #148 with the fix.

Thank you!